TypeCobol/Compiler/Symbols/ProgramSymbol.cs

```diff
@@ -0,0 +1,473 @@
+
+        /// <summary>
+        /// Get the Variable visibility mask.
+        /// </summary>
+        public virtual Flags VariableVisibilityMask => IsNested ? (Flags.Global | Flags.GLOBAL_STORAGE) : 0;
+
+        /// <summary>
+        /// Get the type visibility mask for a Program.
+        /// </summary>
+        public virtual Flags TypeVisibilityMask => IsNested ? (Flags.Global | Flags.Private | Flags.Public) : 0;
+
+        /// <summary>
+        /// Determines if a Type is accessible from this Program.
+        /// </summary>
+        /// <param name="typeSym">The Type To check</param>
+        /// <returns>true if the type is accessible, false otherwise</returns>
+        public virtual bool IsTypeAccessible(TypedefSymbol typeSym)
+        {
+            System.Diagnostics.Debug.Assert(typeSym != null);
+            Symbol typeSymPrg = typeSym.TopParent(Kinds.Program);
+            System.Diagnostics.Debug.Assert(typeSymPrg != null);
+            Symbol myTopPrg = TopParent(Kinds.Program);
+            System.Diagnostics.Debug.Assert(myTopPrg != null);
+            Flags mask = TypeVisibilityMask;
+
+            if (typeSymPrg == myTopPrg)
+            {//Same program ==> Apply the visibility mask
+                return mask == 0 || typeSym.HasFlag(mask);
+            }
+            else
+            {//Different programs ==> only public.
+                return typeSym.HasFlag(Flags.Public);
+            }
+        }
+
+        /// <summary>
+        /// Resolve the given Symbol paths from this scope
+        /// </summary>
+        /// <param name="paths">The qualified path of the symbol reference in COBOL85 order: MUST BE IN LOWER CASES</param>
+        /// <param name="results">The All discovered candidate symbol accumulator</param>
+        /// <param name="bRecurseEnglobingPrograms">true to recurse into englobing variables to look for global variable, false otherwise</param>
+        /// <param name="visibilityMask">Visibility Mask</param>
+        /// <returns>The referenced symbols if any</returns>
+        public Scope<VariableSymbol>.MultiSymbols ResolveReference(string[] paths, Scope<VariableSymbol>.MultiSymbols results, bool bRecurseEnglobingPrograms, Symbol.Flags visibilityMask)
+        {
+            if (paths == null || paths.Length == 0)
+                return results;
+            string name = paths[0];
+            this.Domain.TryGetValue(name, out var candidates);
+            if (candidates != null)
+            {
+                foreach (var candidate in candidates)
+                {
+                    if (visibilityMask != 0)
+                    {
+                        if ((candidate.Flag & visibilityMask) == 0)
+                            continue;
+                    }
+
+                    if (candidate.IsMatchingPath(paths))
+                        results.Add(candidate);
+                }
+            }
+
+            if (results.Count == 0 && bRecurseEnglobingPrograms)
```

Comment:
I don't think you need to test `results.Count == 0` here.
Otherwise this kind of code doesn't works:
```cobol
       IDENTIFICATION DIVISION.
       PROGRAM-ID. DVZFSCT0.
       DATA DIVISION.
       WORKING-STORAGE SECTION.
       01 a pic X global.
      
       IDENTIFICATION DIVISION.
       PROGRAM-ID. Nested.
       DATA DIVISION.
       WORKING-STORAGE SECTION.
       01 a pic X.
       procedure division.
           move a to a
           .
       END PROGRAM Nested.
       END PROGRAM DVZFSCT0.
```
Here `move a to a` must be ambiguous.

Reply:
However i tried to compile this program within RDZ, the compilation was ok, I got only one warning.
Data items"A (ALPHANUMERIC)"and"A (ALPHANUMERIC)"had overlapping storage.  An overlapping move will occur at execution time.
This just mean that we assign a variable to itself, I think.

Reply:
Hum, I think my example was not good. The problem is not the move but rather the symbol resolution of `a`.

The same example but with 2 different variable.
```cobol
       IDENTIFICATION DIVISION.
       PROGRAM-ID. DVZFSCT0.
       DATA DIVISION.
       WORKING-STORAGE SECTION.
       01 a pic X global.
      
       IDENTIFICATION DIVISION.
       PROGRAM-ID. Nested.
       DATA DIVISION.
       WORKING-STORAGE SECTION.
       01 a pic X.
       01 b pic X.
       procedure division.
           move a to b
           .
       END PROGRAM Nested.
       END PROGRAM DVZFSCT0.
```
We except an error because `a` is ambiguous.

Reply:
So it seems this is a TypeCobol Specificity, because I tried again to compile your second sample with Rdz Cobol Compiler, the compilation is Ok.
So Is it a TypeCobol Specificity ?

Reply:
My bad, I thought that variable marked global in parent program were always checked.
So I think we have some bad tests with global because the current variable resolution doesn't works like that.
I've created issue #1360 to fix the bug in the current implementation.

Reply:
Be aware that after discussion with reydelpa and sarterbe we decided that the variable resolution for variable in `global-storage` must not be the same.
See new rule **GLOBALSS_ALWAYS_RESOLVE** in #805.